spatial/r3/transform.go

```diff
@@ -0,0 +1,259 @@
+// Copyright ©2022 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package r3
+
+import (
+	"math"
+)
+
+// Transform represents a 3D spatial transformation.
+// The zero value of Transform is the identity transform.
+type Transform struct {
+	// in order to make the zero value of Transform represent the identity
+	// transform we store it with the identity matrix subtracted.
+	// These diagonal elements are subtracted such that
+	//  d00 = x00-1, d11 = x11-1, d22 = x22-1, d33 = x33-1
+	// where x00, x11, x22, x33 are the matrix diagonal elements.
+	// We can then check for identity in if blocks like so:
+	//  if T == (Transform{})
+	d00, x01, x02, x03 float64
+	x10, d11, x12, x13 float64
+	x20, x21, d22, x23 float64
+	x30, x31, x32, d33 float64
+}
+
+// Apply applies the Transform to the argument vector
+// and returns the result.
+func (t Transform) Apply(v Vec) Vec {
+	if t.isIdentity() {
+		return v
+	}
+	// https://github.com/mrdoob/three.js/blob/dev/src/math/Vector3.js#L262
+	w := 1 / (t.x30*v.X + t.x31*v.Y + t.x32*v.Z + t.d33 + 1)
+	if math.IsInf(w, 0) {
+		return Vec{}
+	}
+	return Vec{
+		X: ((t.d00+1)*v.X + t.x01*v.Y + t.x02*v.Z + t.x03) * w,
+		Y: (t.x10*v.X + (t.d11+1)*v.Y + t.x12*v.Z + t.x13) * w,
+		Z: (t.x20*v.X + t.x21*v.Y + (t.d22+1)*v.Z + t.x23) * w,
+	}
+}
+
+// zeroTransform is the Transform that returns zeroTransform when multiplied by any Transform.
+var zeroTransform = Transform{d00: -1, d11: -1, d22: -1, d33: -1}
+
+// NewTransform returns a new Transform type and populates its elements
+// with values passed in row-major form. If val is nil then NewTransform
+// returns a Transform filled with zeros.
+func NewTransform(a []float64) Transform {
+	if a == nil {
+		return zeroTransform
+	}
+	if len(a) != 16 {
+		panic("Transform is initialized with 16 values")
```

Comment:
This is ambiguous; `panic(fmt.Sprintf("r3: NewTransform invalid slice length: %d != 16", len(a)))`.